src/coreclr/jit/valuenum.cpp

```diff
@@ -7585,7 +7596,16 @@ void Compiler::fgValueNumberTree(GenTree* tree)
                     else
                     {
                         ValueNumPair lclVNPair = varDsc->GetPerSsaData(ssaNum)->m_vnPair;
-                        tree->gtVNPair = vnStore->VNPairApplySelectors(lclVNPair, lclFld->GetFieldSeq(), indType);
+                        if (lclVNPair.GetLiberal() == ValueNumStore::NoVN)
+                        {
+                            // We didn't not assign a correct VN to the local, probably it was written using a different
```

Comment:
@briansull  I am trying to pass the ci, could you please help me with it? The failures look like https://dev.azure.com/dnceng/public/_build/results?buildId=1289712&view=ms.vss-test-web.build-test-results-tab&runId=38073760&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab&resultId=107788
on arm32 windows/linux only with JitStess and only on one test. We assert when try to `ApplySelector` from `NoVN`.
I have added this code for one cases but we have more failures, what I don't understand is why we don't hit these asserts without my changes, do I understand correctly that:
1. LCL_VAR can have a unique VN before the changes;
2. other code parts expect it and do the right things with it;

if both are true why do we hit these asserts and what is the best way to fix it? Should I add similar checks `if (lclVNPair.GetLiberal() == ValueNumStore::NoVN)` to all failing cases?

Reply:
I am trying to understand your issue.  
It is difficult to answer your questions without a full understanding of what is going on here.

Reply:
Generally it is an error to have read with a valenumber of NoVN 

Check if line 7500 (above) applies here:

```
                        // There are a couple of cases where we haven't assigned a valid value number to 'lcl'
                        //
                        if (lcl->gtVNPair.GetLiberal() == ValueNumStore::NoVN)
```

Reply:
Yes, but I don't understand where here do we update `lvaTable[lclNum].GetPerSsaData(lclDefSsaNum)->m_vnPair.SetBoth(lclVarVN);`?
 And when I do `lvaTable[lclNum].GetPerSsaData(lclDefSsaNum)->m_vnPair.SetBoth(ValueNumStore::NoVN);` it starts to produce errors, so what value should I write to `GetPerSsaData` when the local has invalid VN and why don't we write it there in the example on line 7500?

Reply:
@sandreenko 
I can look at a Jit Dump if you have one

Reply:
[JitDumpForBrian.txt](https://github.com/dotnet/runtime/files/6979206/JitDumpForBrian.txt)
this is Jit dump for the assert that VN is not expected to be NoVN.
`000322` is marked as `DontVN` because there is a struct reinterpretation on it:
```
               [000099] --C----------             +--*  BLK       struct<System.Runtime.Intrinsics.Vector64`1[Double], 8>
               [000321] -------------             |  \--*  ADDR      byref
               [000322] $-------N-----            |     \--*  LCL_VAR   struct<System.Runtime.Intrinsics.Vector128`1[Double], 16> V09 tmp3
```
so we later generate a unique VN for it, but VN does not expect `lvaTable[V09].GetPerSsaData()` to be `NoVN` and fails with assert.

Reply:
That is the lhs of an assignment, so it is a write not a read.
I think that lhs don't require/need a ValueNumber

Reply:
I am looking at the dump now

Reply:
It looks like that tree get successfully ValueNumber-ed

```
***** BB03, STMT00023(after)
N005 ( 13, 11) [000365] -A------R----             *  ASG       long   $VN.Void
N004 (  6,  5) [000363] *------N-----             +--*  IND       long   $500
N003 (  3,  3) [000321] -------------             |  \--*  ADDR      byref  Zero Fseq[_00(0xab45adc)] $345
N002 (  3,  2) [000322] $U------N-----            |     \--*  LCL_VAR   struct<System.Runtime.Intrinsics.Vector128`1[Double], 16> V09 tmp3         ud:2->3 $482
N001 (  3,  2) [000364] -------------             \--*  LCL_VAR   long   V15 tmp9         u:2 (last use) $500
```

Reply:
The next tree is where we see the assert:

```
***** BB03, STMT00011(before)
N003 (  7,  5) [000104] -A------R----             *  ASG       struct (copy)
N002 (  3,  2) [000103] D------N-----             +--*  LCL_VAR   struct<System.Runtime.Intrinsics.Vector128`1[Double], 16> V03 loc2         d:2
N001 (  3,  2) [000101] -------------             \--*  LCL_VAR   struct<System.Runtime.Intrinsics.Vector128`1[Double], 16> V09 tmp3         u:3 (last use)

N001 [000101]   LCL_VAR   V09 tmp3         u:3 (last use) => $484 {MemOpaque:NotInLoop}
```

Reply:
In the next tree there is no `GTF_VAR_DONT_VN` set on `[000101] LCL_VAR V09`
So we go down this path:
```
                        else
                        {
                            rhsVNPair = lvaTable[rhsLclVarTree->GetLclNum()]
                                            .GetPerSsaData(rhsLclVarTree->GetSsaNum())
                                            ->m_vnPair;
                            var_types indType = rhsLclVarTree->TypeGet();
                            rhsVNPair = vnStore->VNPairApplySelectors(rhsVNPair, rhsFldSeq, indType);
                        }
```

And read `lvaTable[V09].GetPerSsaData()` which is `NoVN`

Reply:
Perhaps you need to mark the local variable as unsafe to ValueNumber

Reply:
Another fix would be to set 

`lvaTable[V09].GetPerSsaData()  to $500`

when processing  `[000365] `

LCL_VAR   long   V15 tmp9         u:2 (last use) $500

Reply:
If I understand correctly  VN that we print on LHS of an assignment does not matter at all, the real value is saved in `lvaTable[V09].GetPerSsaData() `.
At least it is how I understand this:
https://github.com/dotnet/runtime/blob/7b0a27e370b8b00dccabc11df879a6f59ad75c60/src/coreclr/jit/valuenum.cpp#L7549

but in our case it is a Def, because it is a use:
```
N005 ( 13, 11) [000365] -A------R----             *  ASG       long   $VN.Void
N004 (  6,  5) [000363] *------N-----             +--*  IND       long   $500
N003 (  3,  3) [000321] -------------             |  \--*  ADDR      byref  Zero Fseq[_00(0xab45adc)] $345
N002 (  3,  2) [000322] $U------N-----            |     \--*  LCL_VAR   struct<System.Runtime.Intrinsics.Vector128`1[Double], 16> V09 tmp3         ud:2->3 $482 <- marked as use.
N001 (  3,  2) [000364] -------------             \--*  LCL_VAR   long   V15 tmp9         u:2 (last use) $500
```
but, obviously, we don't care about its value, we won't read it anywhere.

What we care about is value of `lvaTable[V09].GetPerSsaData()` we want it be `NoVN` and we want the next tree that uses it not to try to optimize it or "see" the values of any of the fields inside this lclVar.

Reply:
What you are saying above does sound correct to me.   I haven't made many (or any) changes in this area SSA and VN, so it is also my understanding from reading the code.

Reply:
But with this approach you will have to check for this value for every read (use) site, because the mismatch can occur either at the def or the use.  I the above case we have a mismatch on the def site, which I guess already does write a NoVN, but later at the use site we pull out the NoVN and try to use it in VNPairApplySelectors and the use site won't have the `GTF_VAR_DONT_VN` flag set as the isn't a mismatch on the use site.